### PR TITLE
Generate error message for duplicate items

### LINKED ISSUE #600: Generate a more helpful error message if Compile (or other implicit items) are specified multiple times
If the implicit items are enabled and also explicitly added to the project file, the compiler will generate an error saying that the source file was specified multiple times.  We should detect this situation earlier and provide an error message which tells how to disable the implicit items if desired.

### PR DESCRIPTION
**Pull request by dsplaisted**
Fixes #600

Generate an error message if there are duplicate `Compile`, `EmbeddedResource`, or `Content` items.  The error message looks like this:

> C:\git\dotnet-sdk\bin\Debug\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.DefaultItems.targets(66,5): error : Duplicate Compile items were included. The .NET SDK includes Compile items from your project directory by default. You can either remove these items from your project file, or set the 'EnableDefaultCompileItems' property to 'false' if you want to explicitly include them in your project file. For more information, see https://github.com/dotnet/sdk  The duplicate items were: 'Class1.cs', 'DuplicateCompileItems.cs'

If default items are disabled, then the error message won't include information about how to disable them:

>C:\git\dotnet-sdk\bin\Debug\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.DefaultItems.targets(66,5): error : Duplicate Compile items were included.  The duplicate items were: 'DuplicateCompileItems.cs'

Open questions or other notes:

- Create a fwlink that will point to the documentation for default items.  @blackdwarf, can you do this?
- Should the .NET SDK check for duplicate `Content` items, or since it doesn't have any logic for them by default should that go in the Web SDK?
- This doesn't check for duplicate `None` items, as I don't think those would typically cause an issue
- Does formatting items in a list need to be localized?
- Should we check for duplicate items at all if the default items are disabled?

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/Resources/Strings.resx [hunk 1]
```diff
@@ -225,4 +225,16 @@
...
     <value>Dependency conflict. '{0}' expected '{1}' but received '{2}'</value>
   </data>
+  <data name="DuplicateItemsDefaultExplanation" xml:space="preserve">
+    <value> The .NET SDK includes {0} items from your project directory by default.</value>
+  </data>
+  <data name="DuplicateItemsError" xml:space="preserve">
+    <value>Duplicate {0} items were included.{1}  The duplicate items were: {2}</value>
+  </data>
+  <data name="DuplicateItemsHowToFix" xml:space="preserve">
+    <value> You can either remove these items from your project file, or set the '{0}' property to '{1}' if you want to explicitly include them in your project file.{2}</value>
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.NET.Build.Tasks/Resources/Strings.resx]: I suspect that the experiences of https://github.com/dotnet/sdk/issues/623 and https://github.com/dotnet/roslyn-project-system/issues/1126 indicate that we should also promote the `Update` and `Remove` mechanisms here. But that makes a very long error, which isn't ideal.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets [hunk 2]
```diff
@@ -52,4 +52,45 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+    <CheckForDuplicateItems
+      Items="@(EmbeddedResource)"
+      ItemName="EmbeddedResource"
+      DefaultItemsEnabled="$(EnableDefaultItems)"
+      DefaultItemsOfThisTypeEnabled="$(EnableDefaultEmbeddedResourceItems)"
+      PropertyNameToDisableDefaultItems="EnableDefaultEmbeddedResourceItems"
+      MoreInformationLink="$(DefaultItemsMoreInformationLink)"
+      />
+    
+    <!-- TODO: Do we check for duplicate content items here, or in the Web SDK? EnableDefaultContentItems isn't defined or used at all by the .NET SDK -->
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets]: I'd say to do it here because `Content` is a type known in `Common.targets`, and this is the closest layer above that. If it were defined and used only in Web, I'd say to move it there.

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateItems.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                    defaultExplanation = string.Format(CultureInfo.CurrentCulture, Strings.DuplicateItemsDefaultExplanation, ItemName);
+
+                    defaultExplanation += string.Format(CultureInfo.CurrentCulture, Strings.DuplicateItemsHowToFix,
+                        PropertyNameToDisableDefaultItems,
+                        PropertyValueToDisableDefaultItems,
+                        moreInformation);
+                }
+
+                //  TODO: Does quoting and the separator between items in the list need to be localized?
+                string duplicateItemsFormatted = string.Join(", ", duplicateItems.Select(d => $"'{d.Key}'"));
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateItems.cs]: I don't know what the right answer is. My guess is: change this to a semicolon (like MSBuild uses to separate items) and don't localize it (MSBuild itself doesn't seem to, probably in part because there's no distinction between "format list of items to print" and "format list of items to pass a string to code" which would have different localization needs).

**Review by rainersigwald — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/Resources/Strings.resx (hunk 1 above)
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/Resources/Strings.resx]: This is really hard to follow for translators: leading whitespace, placeholder followed by two spaces followed by another sentence. Placeholder immediately after a sentence. Can we just log consecutive independent messages? If not, maybe don't worry about DRY and just put the full text of each message a user can see so the translator has all of the context. At a minimum, use the `<comment>` element to explain how these compose.

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateItems.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        public string PropertyNameToDisableDefaultItems { get; set; }
+
+        public string PropertyValueToDisableDefaultItems { get; set; } = "false";
+
+        public string MoreInformationLink { get; set; }
+
+        protected override void ExecuteCore()
+        {
+            var duplicateItems = Items.GroupBy(i => i.ItemSpec).Where(g => g.Count() > 1).ToList();
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateItems.cs]: Why ToList()?

**Review by nguerrera — CHANGES_REQUESTED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateItems.cs (hunk 4 above)
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateItems.cs]: Also is g.Count() cheap in practice?

I would be interested in seeing some measurment where this task shows up on the perf summary since it's doing work. Maybe it doesn't register in comparison to surrounding logic, but my instinct from lower level work is to cringe at all of this LINQ.

**Comment by dsplaisted** [comment]
>> Do we check for duplicate content items here, or in the Web SDK? EnableDefaultContentItems isn't defined or used at all by the .NET SDK

> I'd say to do it here because `Content` is a type known in `Common.targets`, and this is the closest layer above that. If it were defined and used only in Web, I'd say to move it there.

I think the question is, are we checking for duplicate items in general, or are we just checking for duplicate items caused by the implicit includes in the SDK?  Does adding an extra check and error message add value if the problem is unrelated to the implicit includes?

I'm leaning towards saying that these checks and error messages should only apply when implicit includes are enabled.  That would also mean that the validation for `Content` items more logically belongs in the Web SDK.

@rainersigwald @nguerrera @srivatsn Thoughts?

**Comment by dsplaisted** [comment]
I've updated the PR to only use one resource string for the error message.  It should now look like this (mostly the same as it was previously):

> C:\git\dotnet-sdk\bin\Debug\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.DefaultItems.targets(66,5): error : Duplicate 'Compile' items were included. The .NET SDK includes 'Compile' items from your project directory by default. You can either remove these items from your project file, or set the 'EnableDefaultCompileItems' property to 'false' if you want to explicitly include them in your project file. The duplicate items were: 'Class1.cs'; 'DuplicateCompileItems.cs' [C:\git\dotnet-sdk\bin\Debug\Tests\It_gives_an_error_message_if_duplicate_compile_items_are_included\DuplicateCompileItems\DuplicateCompileItems\DuplicateCompileItems.csproj]

I've also modified it to only generate the error if the implicit items were enabled.  If not, then it will be up to targets further down the pipeline to detect the issue and generate an error as they always have.

**Comment by dsplaisted** [comment]
@MattGertz for approval

#### Scenario

A user upgrades from RC2 to RC3.  The projects they created with RC2 will generate compile errors saying that the source files were specified multiple times.  This is because we now include the source files implicitly.  This change will generate an error message that explains how to fix the issue:

> error : Duplicate 'Compile' items were included. The .NET SDK includes 'Compile' items from your project directory by default. You can either remove these items from your project file, or set the 'EnableDefaultCompileItems' property to 'false' if you want to explicitly include them in your project file. The duplicate items were: 'Class1.cs'; 'DuplicateCompileItems.cs'

#### Bug

#600

#### Workarounds

n/a

#### Risk

Low - Tests should cover any impacted scenarios

#### Performance Impact

TBD but should be low - it involves calling an additional task for each of the `Compile`, `EmbeddedResource`, and `Content` item types and checking for duplicates.

#### Regression Analysis

n/a

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets [hunk 5]
```diff
@@ -52,4 +52,45 @@ Copyright (c) .NET Foundation. All rights reserved.
     <Content Remove="$(DefaultRemoves)" />
   </ItemGroup>
 
+  <UsingTask TaskName="CheckForDuplicateItems" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
+
+  <Target Name="CheckForDuplicateItems" BeforeTargets="_CheckForInvalidConfigurationAndPlatform">
+
+    <PropertyGroup>
+      <!-- TODO: Create a fwlink for this -->
+      <DefaultItemsMoreInformationLink>https://github.com/dotnet/sdk</DefaultItemsMoreInformationLink>
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets]: We need to have an acceptable link before we merge.

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateItems.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        public string PropertyValueToDisableDefaultItems { get; set; } = "false";
+
+        [Required]
+        public string MoreInformationLink { get; set; }
+
+        protected override void ExecuteCore()
+        {
+            if (DefaultItemsEnabled && DefaultItemsOfThisTypeEnabled)
+            {
+                var duplicateItems = Items.GroupBy(i => i.ItemSpec).Where(g => g.Count() > 1).ToList();
```
Comment (natidea) [inline src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateItems.cs]: Can required inputs be null? Change to 
```csharp
(Items ?? Enumerable<ITaskItem>.Empty())
```

**Review by natidea — APPROVED** [review]

**Comment by natidea** [comment]
Shiproom approved

**Comment by blackdwarf** [comment]
@dsplaisted where does the new link go to?

**Comment by dsplaisted** [comment]
@blackdwarf It goes to https://github.com/dotnet/core/blob/master/release-notes/sdk-implicit-items.md.  You're an owner of the link as well @terrajobst, @richlander, and a few others.

**Comment by AdrienTorris** [comment]
I had this error after upgrading my solution from _Visual Studio 2017 RC_ to _Visual Studio 2017_.
I fix it, I deleted the line `<Compile Include="**\*.cs" />` in all the _csproj_ files of my project.

**Comment by AdrienTorris** [comment]
When I upgraded my project from VS2017 RC to VS2017 it causes an error in all my class library projects. After reinstalling VS (yes I read that it was the solution ...) I deleted the line `<Compile Include="**\*.cs" />` in all the csproj files of my class libraries to can build again.

**Comment by richlander** [comment]
I think this explains it: https://github.com/dotnet/core/blob/master/release-notes/1.0/sdk/1.0-rc3-default-compile-items.md. Does that make sense?
